UI framework and menu additions

Henri-J-Norden opened the pull request:
### Description & Benefits

Framework:
* Add Draw_Checkbox and Redraw_Menu methods
* Draw_Menu_Item supports custom icons if CREALITY_DWIN_EXTUI_CUSTOM_ICONS is defined
* Draw_Menu_Item supports two labels is one item
* Draw_Float can draw NaN
* Status area negative Z value fix
* Draw_Popup keeps selection value for Confirm mode (so selection is preserved when returning to the same menu after confirming)
* Clear_Screen function documentation
* Replace active_menu if-block with switch in Value_Control

Menus:
* Add probe control to Move menu
* Add cold extrude temperature limit and power loss recovery control to Control/Advanced
* Add print statistics to Control/Info
* Refactor Control/Info to use Draw_Menu_Item
* Auto Home moves extruder tip to Z_SAFE_HOMING position when finished
* Fix 'more' argument for multiple menu items, fix typos

### Configurations

Added CREALITY_DWIN_EXTUI_CUSTOM_ICONS option to `Marlin/Configuration.h` and `E3V2 Templates/*`.

### Related Issues
https://github.com/Jyers/Marlin/pull/126
https://github.com/Jyers/Marlin/pull/135

Henri-J-Norden committed:
UI framework changes, menu additions/fixes

Framework:
* Add Draw_Checkbox and Redraw_Menu methods
* Draw_Menu_Item custom icon support
* Draw_Float NaN support
* Status area negative Z value fix
* Draw_Popup keeps selection value for Confirm mode
* Clear_Screen documentation
* Replace active menu if block with switch in Value_Control

Menus:
* Add probe control to Move menu
* Add cold extrude temperature limit and power loss recovery control to Control/Advanced
* Add print statistics to Control/Info
* Auto Home moves extruder tip to Z_SAFE_HOMING position when finished
* Fix 'more' setting for menu items

Henri-J-Norden committed:
Update configuration files to include CREALITY_DWIN_EXTUI_CUSTOM_ICONS

Henri-J-Norden committed:
Display firmware with new icons

Henri-J-Norden committed:
Bugfix

Henri-J-Norden committed:
Add support for dual-row label to Draw_Menu_Item, refactor Info menu to use Draw_Menu_Item

Jyers commented on Marlin/src/lcd/extui/lib/creality_dwin/creality_dwin.cpp:
```diff
@@ -666,13 +704,19 @@ void CrealityDWINClass::Menu_Item_Handler(uint8_t menu, uint8_t item, bool draw/
           else {
             Popup_Handler(Home);
             gcode.process_subcommands_now_P(PSTR("G28"));
+            #if ENABLED(Z_SAFE_HOMING)
+              planner.synchronize();
+              char buf[20];
+              sprintf(buf, "G0 X%i Y%i", Z_SAFE_HOMING_X_POINT, Z_SAFE_HOMING_Y_POINT);
+              gcode.process_subcommands_now_P(buf);
+            #endif
```
Why do you need to send the printer to its homing point here? If Z_SAFE_HOMING is enabled the print head should already be at this position.

Henri-J-Norden replied:
Not exactly: while homing with a bed probe, the extruder is moved such that the _probe_ ends up at (Z_SAFE_HOMING_X_POINT, Z_SAFE_HOMING_Y_POINT) and that's where it leaves it, meaning that the nozzle tip is left at (Z_SAFE_HOMING_X_POINT + NOZZLE_TO_PROBE_OFFSET.x, Z_SAFE_HOMING_Y_POINT + NOZZLE_TO_PROBE_OFFSET.y). This is very annoying for e.g. calibrating the z-offset after auto-homing, as you have to manually move to (Z_SAFE_HOMING_X_POINT, Z_SAFE_HOMING_Y_POINT), otherwise you will be changing the offset based on a point different from the one the Z-axis was homed with (not a big difference if your bed is ~level, but not right either!). In case a bed probe is not used for homing, this should indeed have no effect.

Jyers replied:
That all makes sense, but I'd follow up with this; Why would you be using the auto home item and manually setting your z offset when there is a fully featured z offset menu a few menu items down from that? A menu which will not only handle homing and centering the nozzle, but will move it down to the bed for you and actively move it as you adjust the offset.

Henri-J-Norden replied:
That is a good question, I did use the menu to set the z-offset. I did not, however, use the "Home z-axis" item from that menu - with some further testing right now I found:
1. Maybe an issue: "Home z-axis" does not move to (Z_SAFE_HOMING_X_POINT, Z_SAFE_HOMING_Y_POINT) before homing, unless no homing has been performed and the full auto-home command is issued. This is not necessarily an issue if you deliberately want to set the z-offset based on a point you have manually chosen, but it likely is an issue if the homing point is random (wherever the extruder randomly happened to be positioned).
2. A serious issue: "Home z-axis" suffers from the same issue fixed in the change above - it homes the z-axis, but then leaves the nozzle wherever it ends up. Now you will be setting the z-offset using a point that was not the homing target. Ideally the same fix should be applied to the "Home z-axis" item, or perhaps exclusively to it (but that can cause another positioning inconsistency as in point 1)?

Jyers replied:
Alright, so all things considered I'm going to exclusively fix the z offset auto homing. I'll make sure it always homes at the set homing point, as well as ensuring it bring the nozzle to that point in order to set the offset. I think this should make the changes to the auto home item unnecessary, as its not specifically just for setting the z offset. Thanks for your work, hopefully I should start getting it merged in the next few days.

Jyers committed:
Fixed Custom Icon System

Jyers committed:
Formatting and Menu Fixes

Jyers committed:
Merge branch 'Ender_3_V2_Extensible_UI' into pr/Henri-J-Norden/134

Jyers commented:
I changed the way the custom icons are handled when not enabled to make it a bit cleaner. I just put random substitutions for each, as I don't know what they are being used for. You can either let me know what you think the best substitutions will be or go ahead and commit the new icon id's in creality_dwin.h

Henri-J-Norden committed:
Change custom icon defaults

Henri-J-Norden committed:
Fix strlen(NULL) check in Draw_Menu_Item

Henri-J-Norden commented:
That is indeed a much nicer way of handling the defaults for the custom icons! So far, I've just used ICON_Version for all of them since it's a nice generic-looking one. From the defaults, I removed CUSTOM_ICON_START since it doesn't make sense if there are no custom icons and ICON_Checkbox_F/ICON_Checkbox_T since there is no alternative for them and those values should not be used (Draw_Checkbox should figure out what to do).

I noticed you removed indentation for some of the #if blocks, I had added those where I noticed them based on [Marlin Coding Standards](https://marlinfw.org/docs/development/coding_standards.html), which said: `All block elements should increase the indentation level, including #if blocks and other non-brace compiler blocks`.

Jyers commented:
That all looks good. The indentation was just to keep in matching the rest of the code, but your definitely right, I should probably fix all of the compile time conditional idents to match the marlin standards. That being said I should also probably take a moment to really read through those standards as I have only ever really skimmed through them. Let me know if you discover any other blatant formatting issues that need resolving.

Jyers committed:
The Great Indent

Jyers committed:
Merge branch 'framework' of https://github.com/henri-j-norden/marlin into pr/Henri-J-Norden/134

Jyers committed:
Improved Check Mark

Jyers committed:
Utilize Redraw_Menu and Draw_Checkbox

Jyers committed:
Merge branch 'Ender_3_V2_Extensible_UI' into pr/Henri-J-Norden/134

Jyers committed:
Fix a few checkbox calls

Jyers committed:
Z Offset and Manual Level Fixes

Jonny11bonk commented:
Hi! I changed the color of the icons on my LCD and they dont show up at all for some reason. Any help?

Henri-J-Norden commented:
> Hi! I changed the color of the icons on my LCD and they dont show up at all for some reason. Any help?

Hard to say, without seeing your files. Make sure that they are using the same settings as the original icons.
1. Are you saving as baseline JPEG (NOT progressive)?
2. Are the file sizes around the same? This may be due to some jpeg settings or additional metadata (e.g. added by default when using Photoshop).

Jonny11bonk commented:
> > Hi! I changed the color of the icons on my LCD and they dont show up at all for some reason. Any help?
> 
> Hard to say, without seeing your files. Make sure that they are using the same settings as the original icons.
> 
> 1. Are you saving as baseline JPEG (NOT progressive)?
> 2. Are the file sizes around the same? This may be due to some jpeg settings or additional metadata (e.g. added by default when using Photoshop).

Im using Paint.net. The image file size is almost the same. Some images have even smaller size from the original. I don't have any progressive option while saving. Only the compression 4:2:0, 4:2:2 and 4:2:4. It saves as JPG not JPEG.
